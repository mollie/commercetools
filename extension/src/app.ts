import express, { Application } from 'express';
import handler from '../expressHandler';
import morganMiddleware from './logger/morganMiddleware';

const app: Application = express();

app.use(morganMiddleware); // Logs from http level

app.get('/health', (req, res) => {
  res.send('Ok');
});

app.use(express.json());

app.use('/', handler);

export default app;
