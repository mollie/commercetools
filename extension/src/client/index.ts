import createMollieClient, { MollieClient } from '@mollie/api-client';
import { ClientBuilder, Client, AuthMiddlewareOptions, HttpMiddlewareOptions } from '@commercetools/sdk-client-v2';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}

export function initialiseCommercetoolsClient(): Client {
  const {
    commercetools: { projectKey, clientId, clientSecret, host, authUrl, scopes },
  } = config;

  const authMiddlewareOptions: AuthMiddlewareOptions = {
    host: authUrl,
    projectKey,
    credentials: {
      clientId,
      clientSecret,
    },
    scopes,
    fetch,
  };

  const httpMiddlewareOptions: HttpMiddlewareOptions = {
    host,
    fetch,
  };

  const client: Client = new ClientBuilder()
    .withClientCredentialsFlow(authMiddlewareOptions)
    .withHttpMiddleware(httpMiddlewareOptions)
    .withUserAgentMiddleware()
    .build();

  return client;
}
