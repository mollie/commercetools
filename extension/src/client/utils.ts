import createMollieClient, { MollieClient } from '@mollie/api-client';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}
