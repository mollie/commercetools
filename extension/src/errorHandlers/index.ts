import { formatMollieErrorResponse } from './formatMollieErrorResponse';
import { formatExtensionErrorResponse } from './formatExtensionErrorResponse';

const errorHandler = {
  mollie: formatMollieErrorResponse,
  extension: formatExtensionErrorResponse,
};

export default errorHandler;
