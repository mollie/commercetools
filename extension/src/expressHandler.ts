import {Request, Response} from 'express';
import handleRequest from './requestHandlers/handleRequest';
import {HandleRequestInput, HandleRequestSuccess} from './types';
import actions from "./requestHandlers/actions";


export default async function handler(req: Request, res: Response) {
    const requestInput = new HandleRequestInput(req.path, req.method, req.body);
    const result = await handleRequest(requestInput);
    if (result instanceof HandleRequestSuccess) {
        return res.status(200).send({actions: actions});

    } else {
        return res.status(result.status).send({errors: result.errors});
    }
};
