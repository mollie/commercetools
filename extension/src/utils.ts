import { MethodsListParams } from '@mollie/api-client';
/**
 * Generates an ISO string date
 * @returns {String} Returns the current date converted to ISO.
 */
export function createDateNowString() {
  return new Date().toISOString();
}

export function amountMapper(amountPlanned: any): string {
  const { centAmount, fractionDigits } = amountPlanned;
  const divider = Math.pow(10, fractionDigits || 2);
  const mollieAmount = (centAmount / divider).toFixed(2);
  return mollieAmount
}

export function methodListMapper(ctObj: any): MethodsListParams {
  // Generally this shouldn't be needed, but a safety anyway.. eventually could return error here
  if (!ctObj.amountPlanned) {
    return {};
  }
  const mollieAmount = amountMapper(ctObj.amountPlanned);

  const mObject: MethodsListParams = {
    amount: {
      value: mollieAmount,
      currency: ctObj.amountPlanned.currencyCode,
    },
    // Resource is hardcoded, for the time being we only support Orders API
    resource: 'orders',
  };

  if (ctObj.custom?.fields?.paymentMethodsRequest) {
    const parsedMethodsRequest = JSON.parse(ctObj.custom?.fields?.paymentMethodsRequest)
    const { locale, billingCountry, includeWallets, orderLineCategories, issuers, pricing, sequenceType } = parsedMethodsRequest;
    const include = issuers || pricing ? `${issuers ? 'issuers,' : ''}${pricing ? 'pricing' : ''}` : undefined;

    Object.assign(
      mObject,
      locale && { locale: locale },
      include && { include: include },
      includeWallets && { includeWallets: includeWallets },
      billingCountry && { billingCountry: billingCountry },
      sequenceType && { sequenceType: sequenceType },
      orderLineCategories && { orderLineCategories: orderLineCategories },
    );
  }

  return mObject;
}
