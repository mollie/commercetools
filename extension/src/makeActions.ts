import { time } from 'console';
import { v4 as uuid } from 'uuid';
import { Action, ControllerAction, CTTransactionState } from './types';
import { createDateNowString } from './utils';

/**
 *
 * @param customFieldName
 * @param customFieldValue
 * If the customFieldValue is an API response, JSON Stringify it before passing it
 */
const setCustomField = (customFieldName: string, customFieldValue: string) => {
  return {
    action: 'setCustomField',
    name: customFieldName,
    value: customFieldValue,
  };
};

type createInterfaceInteractionParams = {
  actionType: ControllerAction;
  requestValue: string;
  responseValue: string;
  id?: string;
  timestamp?: string;
};
/**
 * @param parameters type createInterfaceInteractionParams, which contains
 * actionType ControllerAction
 * requestValue string
 * responseValue string
 * id string string, optional
 * timestamp string, optional
 * If the responseValue is an API response, JSON Stringify it before passing it
 */
const addInterfaceInteraction = (params: createInterfaceInteractionParams) => {
  const { actionType, requestValue, responseValue, id, timestamp } = params;
  const interfaceInteractionId = id ? id : uuid();
  const interfaceInteractionTimestamp = timestamp ? timestamp : createDateNowString();
  return {
    action: 'addInterfaceInteraction',
    type: {
      key: 'ct-mollie-integration-interface-interaction-type',
    },
    fields: {
      id: interfaceInteractionId,
      actionType,
      createdAt: interfaceInteractionTimestamp,
      request: requestValue,
      response: responseValue,
    },
  };
};

/**
 *
 * @param key will be the mollie order id
 */
const setKey = (key: string) => {
  return {
    action: 'setKey',
    key,
  };
};

/**
 *
 * @param interfaceText will be the mollie order status
 */
const setStatusInterfaceText = (interfaceText: string) => {
  return {
    action: 'setMethodInfoName',
    interfaceText,
  };
};

/**
 *
 * @param id transaction to be updated
 * @param newState CTTransactionState
 */
const changeTransactionState = (id: string, newState: CTTransactionState) => {
  return {
    action: 'changeTransactionState',
    transactionId: id,
    state: newState,
  };
};

/**
 *
 * @param id transaction to be updated
 * @param interactionId either the mollie payment id, or the corresponding interfaceInteraction's id
 */
const changeTransactionInteractionId = (id: string, interactionId: string) => {
  return {
    action: 'changeTransactionInteractionId',
    transactionId: id,
    interactionId,
  };
};

/**
 *
 * @param id transaction to be updated
 * @param timestamp CT DateTime is a JSON string representation of UTC date & time in ISO 8601 format (YYYY-MM-DDThh:mm:ss.sssZ)
 * for example: "2018-10-12T14:00:00.000Z"
 */
const changeTransactionTimestamp = (id: string, timestamp: string) => {
  return {
    action: 'changeTransactionTimestamp',
    transactionId: id,
    timestamp,
  };
};

export const makeActions = {
  setCustomField,
  addInterfaceInteraction,
  setKey,
  setStatusInterfaceText,
  changeTransactionState,
  changeTransactionInteractionId,
  changeTransactionTimestamp,
};
