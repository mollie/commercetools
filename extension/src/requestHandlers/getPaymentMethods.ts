import { MollieClient, List, Method } from '@mollie/api-client';
import { Request } from 'express';
import { CTUpdatesRequestedResponse, Action } from '../types';
import { createDateNowString } from '../utils';

export default async function getPaymentMethods(req: Request, mollieClient: MollieClient): Promise<CTUpdatesRequestedResponse> {
  try {
    const methods: List<Method> = await mollieClient.methods.all();
    const availablePaymentMethods: string = methods.length > 0 ? JSON.stringify(methods) : 'NO_AVAILABLE_PAYMENT_METHODS';
    const ctUpdateActions: Action[] = [
      {
        action: 'addInterfaceInteraction',
        type: {
          key: 'ct-mollie-integration-interface-interaction-type',
        },
        fields: {
          actionType: 'getPaymentMethods',
          request: JSON.stringify(req.body?.custom?.fields?.paymentMethodsRequest),
          response: availablePaymentMethods,
          createdAt: createDateNowString(),
        },
      },
      {
        action: 'setCustomField',
        name: 'paymentMethodsResponse',
        value: availablePaymentMethods,
      },
    ];
    return {
      actions: ctUpdateActions,
      status: 200,
    } as CTUpdatesRequestedResponse;
  } catch (error: any) {
    console.warn(error);
    return error;
  }
}
