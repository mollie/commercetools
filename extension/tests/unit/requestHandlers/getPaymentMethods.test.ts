import { Request } from 'express';
import { mocked } from 'ts-jest/utils';
import getPaymentMethods from '../../../src/requestHandlers/getPaymentMethods';
import { createDateNowString } from '../../../src/utils';
import Logger from '../../../src/logger/logger';

jest.mock('../../../src/utils');

describe('getPaymentMethods unit tests', () => {
  const mockLogError = jest.fn();
  beforeAll(() => {
    Logger.error = mockLogError;
    mocked(createDateNowString).mockReturnValue('2021-10-08T12:12:02.625Z');
  });
  afterAll(() => {
    jest.clearAllMocks();
  });

  it('Should call mollie mollieClient.methods.list', async () => {
    const mockedRequest = {
      custom: { fields: { paymentMethodsRequest: {} } },
    };
    const mollieClient = {
      methods: { list: jest.fn().mockResolvedValueOnce([]) },
    } as any;
    await getPaymentMethods(mockedRequest, mollieClient);
    expect(mollieClient.methods.list).toBeCalled();
  });

  it('Should return status and one update action for commercetools', async () => {
    const mockedPaymentMethodsRequest = '{"locale":"en_US","resource":"orders","billingCountry":"NL","includeWallets":"applepay","orderLineCategories":"eco,meal"}';
    const mockedRequest = {
      custom: {
        fields: { paymentMethodsRequest: mockedPaymentMethodsRequest },
      },
    };
    const mockedMethodsResponse = [
      {
        resource: 'method',
        id: 'ideal',
        description: 'iDEAL',
        minimumAmount: { value: '0.01', currency: 'EUR' },
        maximumAmount: { value: '50000.00', currency: 'EUR' },
        image: {
          size1x: 'https://www.mollie.com/external/icons/payment-methods/ideal.png',
          size2x: 'https://www.mollie.com/external/icons/payment-methods/ideal%402x.png',
          svg: 'https://www.mollie.com/external/icons/payment-methods/ideal.svg',
        },
      },
      {
        resource: 'method',
        id: 'paypal',
        description: 'PayPal',
        minimumAmount: { value: '0.01', currency: 'EUR' },
        maximumAmount: null,
        image: {
          size1x: 'https://www.mollie.com/external/icons/payment-methods/paypal.png',
          size2x: 'https://www.mollie.com/external/icons/payment-methods/paypal%402x.png',
          svg: 'https://www.mollie.com/external/icons/payment-methods/paypal.svg',
        },
      },
    ] as any;
    mockedMethodsResponse.count = 2;
    const mollieClient = {
      methods: { list: jest.fn().mockResolvedValueOnce(mockedMethodsResponse) },
    } as any;
    const { actions, status } = await getPaymentMethods(mockedRequest, mollieClient);

    expect(status).toBe(200);
    expect(actions).toHaveLength(1);
    actions?.forEach(action => {
      expect(action).toMatchSnapshot();
    });
  });

  it('Should return NO_PAYMENT_METHODS when methods returned are empty', async () => {
    const mockedPaymentMethodsRequest = '{"locale":"en_US","resource":"orders","billingCountry":"NL","includeWallets":"applepay","orderLineCategories":"eco,meal"}';
    const mockedRequest = {
      custom: {
        fields: { paymentMethodsRequest: mockedPaymentMethodsRequest },
      },
    };
    const mollieClient = {
      methods: {
        list: jest.fn().mockResolvedValueOnce([
          {
            count: 0,
            links: {
              documentation: {
                href: 'https://docs.mollie.com/reference/v2/methods-api/list-methods',
                type: 'text/html',
              },
              self: {
                href: 'https://api.mollie.com/v2/methods',
                type: 'application/hal+json',
              },
            },
            nextPage: undefined,
            nextPageCursor: undefined,
            previousPage: undefined,
            previousPageCursor: undefined,
          },
        ]),
      },
    } as any;
    const { actions, status } = await getPaymentMethods(mockedRequest, mollieClient);

    expect(status).toBe(200);
    expect(actions).toHaveLength(1);
    actions?.forEach(action => {
      expect(action).toMatchSnapshot();
    });

    const paymentMethodsResponseCTCustomField = actions?.find(a => a.action === 'setCustomField');
    expect(paymentMethodsResponseCTCustomField?.value).toEqual(JSON.stringify({ count: 0, methods: 'NO_AVAILABLE_PAYMENT_METHODS' }));
  });

  it('Should return error if mollieClient call fails', async () => {
    const mockedError = new Error('Test error');
    const mockedRequest = {} as Request;
    const mollieClient = {
      methods: { list: jest.fn().mockRejectedValue(mockedError) },
    } as any;
    const { errors, status } = await getPaymentMethods(mockedRequest, mollieClient);
    expect(status).toBe(400);
    expect(errors).toBeInstanceOf(Array);
    expect(mockLogError).toHaveBeenCalledTimes(1);
  });
});
