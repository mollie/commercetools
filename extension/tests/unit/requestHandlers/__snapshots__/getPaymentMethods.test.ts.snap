// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPaymentMethods unit tests Should return NO_PAYMENT_METHODS when methods returned are empty 1`] = `
Object {
  "action": "addInterfaceInteraction",
  "fields": Object {
    "actionType": "getPaymentMethods",
    "createdAt": "2021-10-08T12:12:02.625Z",
    "request": "{\\"locale\\":\\"en_US\\",\\"resource\\":\\"orders\\",\\"billingCountry\\":\\"NL\\",\\"includeWallets\\":\\"applepay\\",\\"orderLineCategories\\":\\"eco,meal\\"}",
    "response": "{\\"count\\":0,\\"methods\\":\\"NO_AVAILABLE_PAYMENT_METHODS\\"}",
  },
  "type": Object {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`getPaymentMethods unit tests Should return NO_PAYMENT_METHODS when methods returned are empty 2`] = `
Object {
  "action": "setCustomField",
  "name": "paymentMethodsResponse",
  "value": "{\\"count\\":0,\\"methods\\":\\"NO_AVAILABLE_PAYMENT_METHODS\\"}",
}
`;

exports[`getPaymentMethods unit tests Should return status and one update action for commercetools 1`] = `
Object {
  "action": "setCustomField",
  "name": "paymentMethodsResponse",
  "value": "{\\"count\\":2,\\"methods\\":[{\\"resource\\":\\"method\\",\\"id\\":\\"ideal\\",\\"description\\":\\"iDEAL\\",\\"minimumAmount\\":{\\"value\\":\\"0.01\\",\\"currency\\":\\"EUR\\"},\\"maximumAmount\\":{\\"value\\":\\"50000.00\\",\\"currency\\":\\"EUR\\"},\\"image\\":{\\"size1x\\":\\"https://www.mollie.com/external/icons/payment-methods/ideal.png\\",\\"size2x\\":\\"https://www.mollie.com/external/icons/payment-methods/ideal%402x.png\\",\\"svg\\":\\"https://www.mollie.com/external/icons/payment-methods/ideal.svg\\"}},{\\"resource\\":\\"method\\",\\"id\\":\\"paypal\\",\\"description\\":\\"PayPal\\",\\"minimumAmount\\":{\\"value\\":\\"0.01\\",\\"currency\\":\\"EUR\\"},\\"maximumAmount\\":null,\\"image\\":{\\"size1x\\":\\"https://www.mollie.com/external/icons/payment-methods/paypal.png\\",\\"size2x\\":\\"https://www.mollie.com/external/icons/payment-methods/paypal%402x.png\\",\\"svg\\":\\"https://www.mollie.com/external/icons/payment-methods/paypal.svg\\"}}]}",
}
`;

exports[`getPaymentMethods unit tests Should return status and two update actions for commercetools 1`] = `
Object {
  "action": "addInterfaceInteraction",
  "fields": Object {
    "actionType": "getPaymentMethods",
    "createdAt": "2021-10-08T12:12:02.625Z",
    "request": "{\\"locale\\":\\"en_US\\",\\"resource\\":\\"orders\\",\\"billingCountry\\":\\"NL\\",\\"includeWallets\\":\\"applepay\\",\\"orderLineCategories\\":\\"eco,meal\\"}",
    "response": "{\\"count\\":2,\\"methods\\":[{\\"resource\\":\\"method\\",\\"id\\":\\"ideal\\",\\"description\\":\\"iDEAL\\",\\"minimumAmount\\":{\\"value\\":\\"0.01\\",\\"currency\\":\\"EUR\\"},\\"maximumAmount\\":{\\"value\\":\\"50000.00\\",\\"currency\\":\\"EUR\\"},\\"image\\":{\\"size1x\\":\\"https://www.mollie.com/external/icons/payment-methods/ideal.png\\",\\"size2x\\":\\"https://www.mollie.com/external/icons/payment-methods/ideal%402x.png\\",\\"svg\\":\\"https://www.mollie.com/external/icons/payment-methods/ideal.svg\\"}},{\\"resource\\":\\"method\\",\\"id\\":\\"paypal\\",\\"description\\":\\"PayPal\\",\\"minimumAmount\\":{\\"value\\":\\"0.01\\",\\"currency\\":\\"EUR\\"},\\"maximumAmount\\":null,\\"image\\":{\\"size1x\\":\\"https://www.mollie.com/external/icons/payment-methods/paypal.png\\",\\"size2x\\":\\"https://www.mollie.com/external/icons/payment-methods/paypal%402x.png\\",\\"svg\\":\\"https://www.mollie.com/external/icons/payment-methods/paypal.svg\\"}}]}",
  },
  "type": Object {
    "key": "ct-mollie-integration-interface-interaction-type",
  },
}
`;

exports[`getPaymentMethods unit tests Should return status and two update actions for commercetools 2`] = `
Object {
  "action": "setCustomField",
  "name": "paymentMethodsResponse",
  "value": "{\\"count\\":2,\\"methods\\":[{\\"resource\\":\\"method\\",\\"id\\":\\"ideal\\",\\"description\\":\\"iDEAL\\",\\"minimumAmount\\":{\\"value\\":\\"0.01\\",\\"currency\\":\\"EUR\\"},\\"maximumAmount\\":{\\"value\\":\\"50000.00\\",\\"currency\\":\\"EUR\\"},\\"image\\":{\\"size1x\\":\\"https://www.mollie.com/external/icons/payment-methods/ideal.png\\",\\"size2x\\":\\"https://www.mollie.com/external/icons/payment-methods/ideal%402x.png\\",\\"svg\\":\\"https://www.mollie.com/external/icons/payment-methods/ideal.svg\\"}},{\\"resource\\":\\"method\\",\\"id\\":\\"paypal\\",\\"description\\":\\"PayPal\\",\\"minimumAmount\\":{\\"value\\":\\"0.01\\",\\"currency\\":\\"EUR\\"},\\"maximumAmount\\":null,\\"image\\":{\\"size1x\\":\\"https://www.mollie.com/external/icons/payment-methods/paypal.png\\",\\"size2x\\":\\"https://www.mollie.com/external/icons/payment-methods/paypal%402x.png\\",\\"svg\\":\\"https://www.mollie.com/external/icons/payment-methods/paypal.svg\\"}}]}",
}
`;
